compiler/src/iree/compiler/Dialect/Flow/Transforms/PrintDispatchGraph.cpp

```diff
@@ -0,0 +1,438 @@
+      Optional<Node> prevNode;
+      for (Operation &op : block) {
+        Node nextNode = processOperation(&op);
+        if (printControlFlowEdges && prevNode)
+          emitEdgeStmt(*prevNode, nextNode, /*label=*/"",
+                       kLineStyleControlFlow);
+        prevNode = nextNode;
+      }
+    });
+  }
+
+  bool isScalarConstantOp(Operation *op) {
+    if (auto constOp = dyn_cast<mlir::arith::ConstantOp>(op))
+      if (constOp.getResult().getType().isIntOrIndexOrFloat()) return true;
+
+    return false;
+  }
+
+  /// Process an operation. If the operation has regions, emit a cluster.
+  /// Otherwise, emit a node.
+  Node processOperation(Operation *op) {
+    Node node;
+
+    if (isScalarConstantOp(op)) {
+      // don't handle scalar constant because it creates too many edges
+      // from a single constant.
+      return node;
+    }
+
+    if (op->getNumRegions() == 1) {
+      // do not generate a cluster when there is one region.
+      processRegion(op->getRegion(0));
+    } else if (op->getNumRegions() > 0) {
+      // Emit cluster for op with regions.
+      node = emitClusterStmt(
+          [&]() {
+            for (Region &region : op->getRegions()) processRegion(region);
+          },
+          getLabel(op));
+    } else {
+      node = emitNodeStmt(op);
+    }
+
+    // Insert data flow edges originating from each operand.
+    if (printDataFlowEdges) {
+      unsigned numOperands = op->getNumOperands();
+      for (unsigned i = 0; i < numOperands; i++) {
+        auto operand = op->getOperand(i);
+
+        // a constant operand is not going to be available in the map.
+        if (valueToNode.count(operand)) {
+          emitEdgeStmt(valueToNode[op->getOperand(i)], node,
+                       /*label=*/numOperands == 1 ? "" : std::to_string(i),
+                       kLineStyleDataFlow);
+        }
+      }
+    }
+
+    for (Value result : op->getResults()) valueToNode[result] = node;
+
+    return node;
+  }
+
+  /// Process a region.
+  void processRegion(Region &region) {
+    for (Block &block : region.getBlocks()) processBlock(block);
+  }
+
+  /// Truncate long strings.
+  std::string truncateString(std::string str) {
+    if (str.length() <= maxLabelLen) return str;
+    return str.substr(0, maxLabelLen) + "...";
+  }
+
+  /// Output stream to write DOT file to.
+  raw_indented_ostream os;
+  /// A list of edges. For simplicity, should be emitted after all nodes were
+  /// emitted.
+  std::vector<std::string> edges;
+  /// Mapping of SSA values to Graphviz nodes/clusters.
+  DenseMap<Value, Node> valueToNode;
+  /// Counter for generating unique node/subgraph identifiers.
+  int counter = 0;
+};
+
+}  // namespace
+
+std::unique_ptr<Pass> createPrintDispatchGraphPass(raw_ostream &os) {
+  return std::make_unique<PrintOpPass>(os);
+}
+
+/// Generate a CFG for a region and show it in a window.
+static void llvmViewGraph(Region &region, const Twine &name) {
```

Comment:
unused? (copied from somewhere else?)

Reply:
Yes, the base code is ViewOpGraph.cpp. I will remove it.